### ISSUE TITLE
Edited message on telegram creates infinite loop

### ISSUE THREAD

**Issue by benos** [RasaHQ/rasa#5654]
**Rasa version**: 1.9.5

**Rasa SDK version** (if used & relevant): n/a

**Rasa X version** (if used & relevant): n/a

**Python version**: unsure. using the Docker image rasa/rasa:1.9.5-full

**Operating system** (windows, osx, ...): Docker on OSX, using rasa/rasa:1.9.5-full

**Issue**:

When a user edits a message on Telegram, this creates a 500 error on rasa, because the payload contains "edited_message" rather than "message" (see screenshots).

This then creates an infinite loop because Rasa sends a 500 back so Telegram keeps retrying. And Telegram is *very* persistent.

Where it gets worse is that Telegram seems to queue all subsequent messages for retry, even if those went through OK. So this can all get pretty apocalyptic (I was only able to get back control by creating a fake endpoint that just sent a bunch of 200 back to Telegram to empty the queue).

I reckon it's the same problem as described here: https://github.com/python-telegram-bot/python-telegram-bot/issues/1244 The proposed solution there is to use "effective_message", which apparently would cover both "message" and "edited_message".

**Error (including full traceback)**:
![image](https://user-images.githubusercontent.com/1001662/79471188-92d5da80-8002-11ea-9b57-983281dfe0ad.png)

![image](https://user-images.githubusercontent.com/1001662/79471333-c7499680-8002-11ea-94df-3e3f0075c06e.png)

```
Traceback (most recent call last):
  File "/opt/venv/lib/python3.6/site-packages/sanic/app.py", line 976, in handle_request
    response = await response
  File "/usr/local/lib/python3.6/asyncio/coroutines.py", line 110, in __next__
    return self.gen.send(None)
  File "/opt/venv/lib/python3.6/site-packages/rasa/core/channels/telegram.py", line 215, in message
    if self._is_user_message(msg):
  File "/opt/venv/lib/python3.6/site-packages/rasa/core/channels/telegram.py", line 175, in _is_user_message
    return message.text is not None
AttributeError: 'NoneType' object has no attribute 'text'
```

**Command or request that led to error**:

![image](https://user-images.githubusercontent.com/1001662/79471366-d3355880-8002-11ea-9c71-d323bee497ba.png)

(this is on Telegram for Desktop (Mac OS X). Just type a message, send it, arrow up to edit, make a change and resend)

**Content of configuration file (config.yml)** (if relevant):
Not relevant

**Content of domain file (domain.yml)** (if relevant):
Not relevant

**Comment by sara-tagger** [comment]
Thanks for raising this issue, @JustinaPetr will get back to you about it soon✨

###### Please also check out the [docs](https://rasa.com/docs/) and the [forum](https://forum.rasa.com/) in case your issue was raised there too 🤗

**Comment by digitalWestie** [comment]
I can confirm this is an issue on 1.9.6. 

As far as I can see from here: https://github.com/RasaHQ/rasa/blob/1.9.6/rasa/core/channels/telegram.py#L214

Shortly after a check for whether the webhook request was a button message, the rest of the code proceeds assuming there is a "message"  key in the post data. 

This won't be the case for a number of different events e.g. 

- edited_message
- edited_channel_post
- channel_post

There needs to be more telegram message type checks to handle responses appropriately. For the meantime, it would be good to catch this possibility and display a warning or something? If I have a moment I will submit a PR to that effect.

**Comment by benos** [comment]
@JustinaPetr Was wondering if you had a chance to look into this

**Comment by erohmensing** [comment]
Hey @benos, sorry you haven't gotten a response yet. I'm wondering how you expect a bot to handle edited messages? As soon as you send a message, rasa will receive it and try to respond to it. If you edit it, do you expect Rasa to assume the old message as invalid and forget it, and process the new one? If so, what does that look like on the user side when the bot's invalid response is still in the UI before the newly valid one?

**Comment by erohmensing** [comment]
> This then creates an infinite loop because Rasa sends a 500 back so Telegram keeps retrying. And Telegram is very persistent.

> Where it gets worse is that Telegram seems to queue all subsequent messages for retry, even if those went through OK. So this can all get pretty apocalyptic (I was only able to get back control by creating a fake endpoint that just sent a bunch of 200 back to Telegram to empty the queue).

I agree that this sounds bad though 😅Ideally we'd just ignore the edit or something here

**Comment by benos** [comment]
Hi @erohmensing , 

Thanks for looking into this. I don't think ignoring the first message would work for reasons you describe, plus the fact that it might already have been passed on from Rasa further down the stack, so that would need cancelling too, which seems really impractical.

So I see two options:

1. Treat the edited message as a new one. It appears this would be a fairly simple change, namely use a different property to get the content (effective_message). See https://github.com/python-telegram-bot/python-telegram-bot/issues/1244 (note that I haven't tested this myself).

2. If it is possible to identify it as the same message that was previously sent, then just say thank you to Telegram (by returning a 200) and ignore the new message.

I've been having another bug where Telegram frequently resends messages that were the product of a button tap (still haven't worked out why - no errors nor pattern I can see), so my vote would be for #2 since it would kill that bug too. And I think it would have the fewer side effects from the user's point of view.

But #1 would be a significant improvement to the current situation.

Thanks,

Ben

**Comment by benos** [comment]
I suppose you could also do #2 just by checking for the edited_message property

**Comment by erohmensing** [comment]
Yeah my original instinct would just be to ignore messages with the `edited_message` property. I'm not sure if treating the edited message as a new one is what people would expect the behavior to look like 🤔

Should we go forward with ignoring the edits causing the infinite loop (and maybe figuring out the button tap thing on the way) and discussed edited message UX further? I think that's a more complicated thing that may just be out of scope (it's a common conversational problem on par of how to respond to multiple messages from a user) 

Would you be interested in contributing a bug fix?

**Comment by benos** [comment]
Sounds good. 

I'm afraid my Python is way to weak to contribute a bug fix (we're only configuring Rasa - the rest of our Stack is on node)

**Comment by erohmensing** [comment]
No worries. Let me take a look into it!

**Comment by benos** [comment]
Hi @erohmensing 

To expand on this a bit:

The button thing is becoming a daily occurrence and renders the combination of Telegram, Rasa and buttons unusable in production and borderline even for testing.

Here is an example of the occurrence, where the user pressed "all right" only once. Unfortunately I can't see any error or condition that would explain why it happens.

![bug](https://user-images.githubusercontent.com/1001662/83861400-d6a2b180-a720-11ea-9002-50d1c444c41b.png)

Also note that other people are reporting the original issue (see 5870)

Should I create another bug report for the buttons or leave it here since the two problems might have the same resolution (ignore duplicate messages)?

Thanks in advance,

Ben

**Comment by benos** [comment]
Also this post on the forum describes a similar issue:

https://forum.rasa.com/t/problem-with-multiple-user-using-telegram/27273

**Comment by benos** [comment]
hi @erohmensing,

I've been looking at ways to work around this issue. A couple of things I found:

- You can get a ID for the message (which is different from the update ID) in the Telegram channel code using this: message_id = msg.message_id
- This can then get passed on to the tracker when calling UserMessage(...)
- I then worked around the issue on my action server by checking for duplicate message_ids in the tracker. It would be better if these were filtered out by rasa immediately

Telegram also seems to expect an immediate acknowledgement for button presses (you can see the "clock" icon on the button staying for a really long time when using it with Rasa). The solution for that appears to be the line in bold

if self._is_button(update):
                    msg = update.callback_query.message
                    text = update.callback_query.data
                    **update.callback_query.answer()** 

Hope this helps
